truffleruby/src/main/ruby/core/truffle/cext.rb

```diff
@@ -958,7 +958,22 @@ def rb_funcallv_public(recv, meth, args)
   end
 
   def rb_funcallv(recv, meth, args)
-    recv.send(meth, *args)
+    rb_funcall(recv, meth, *args)
+  end
+
+  def rb_funcall(recv, meth, *args)
+    old_c_block = Thread.current[:__C_BLOCK__]
+    begin
+      block = Thread.current[:__C_BLOCK__]
+      Thread.current[:__C_BLOCK__] = nil
+      if block
+        recv.send(meth, *args, &block)
+      else
+        recv.send(meth, *args)
```

Comment:
If block is only a Proc or `nil` then you can pass it with `&block` just as well.